components/director/internal/tenantmapping/handler.go

```diff
@@ -1,107 +1,143 @@
-	if request.Method != http.MethodPost {
-		http.Error(writer, fmt.Sprintf("Bad request method. Got %s, expected POST", request.Method), http.StatusBadRequest)
+func (h *Handler) ServeHTTP(writer http.ResponseWriter, req *http.Request) {
+	if req.Method != http.MethodPost {
+		http.Error(writer, fmt.Sprintf("Bad request method. Got %s, expected POST", req.Method), http.StatusBadRequest)
 		return
 	}
 
-	logBuilder := strings.Builder{}
-	logBuilder.WriteString(fmt.Sprintf("\nHeaders: %+v", request.Header))
-
-	writer.Header().Set("Content-Type", "application/json")
-
-	var data Data
-	err := json.NewDecoder(request.Body).Decode(&data)
+	reqData, err := parseRequest(req)
 	if err != nil {
-		if err == io.EOF {
-			http.Error(writer, "Request body is empty", http.StatusBadRequest)
-			return
-		}
-
-		wrappedErr := errors.Wrap(err, "while decoding request body")
-		http.Error(writer, wrappedErr.Error(), http.StatusInternalServerError)
+		respondWithError(writer, http.StatusBadRequest, err, "while parsing the request")
 		return
 	}
 
-	defer func() {
-		err := request.Body.Close()
-		if err != nil {
-			wrappedErr := errors.Wrap(err, "while decoding request body")
-			log.Error(wrappedErr)
-			http.Error(writer, wrappedErr.Error(), http.StatusInternalServerError)
-		}
-	}()
+	tx, err := h.transact.Begin()
+	if err != nil {
+		respondWithError(writer, http.StatusInternalServerError, err, "while opening the db transaction")
+		return
+	}
+	defer h.transact.RollbackUnlessCommited(tx)
 
-	logBuilder.WriteString(fmt.Sprintf("\nInput: %+v", data))
+	ctx := persistence.SaveToContext(context.TODO(), tx)
 
-	if data.Extra == nil {
-		data.Extra = make(map[string]interface{})
+	credID, tenantID, refObjectType, err := h.lookupForTenant(ctx, reqData)
+	if err != nil {
+		respondWithError(writer, http.StatusInternalServerError, err, "while looking for tenant")
+		return
 	}
 
-	extraMap, ok := data.Extra.(map[string]interface{})
-	if !ok {
-		err := fmt.Errorf("Incorrect type %T; expected map[string]interface{}\n", data.Extra)
-		log.Info(logBuilder.String())
-		log.Error(err)
-		http.Error(writer, err.Error(), http.StatusInternalServerError)
+	reqData.Extra["tenant"] = tenantID
+
+	scope, err := h.lookupForScope(credID, refObjectType)
+	if err != nil {
+		respondWithError(writer, http.StatusInternalServerError, err, "while looking for scopes")
 		return
 	}
 
-	extraMap["tenant"] = "9ac609e1-7487-4aa6-b600-0904b272b11f"
+	reqData.Extra["scope"] = scope
 
-	_, ok = extraMap["scope"]
-	if !ok {
-		h.setScopes(extraMap)
+	writer.Header().Set("Content-Type", "application/json")
+	err = json.NewEncoder(writer).Encode(reqData)
+	if err != nil {
+		respondWithError(writer, http.StatusInternalServerError, err, "while encoding data")
+		return
 	}
+}
 
-	data.Extra = extraMap
+func (h *Handler) lookupForTenant(ctx context.Context, reqData ReqData) (string, string, CredenialRefObjectType, error) {
```

Comment:
I don't like the idea of 4 return arguments - it's simply too much, especially if you have equal types (like here: string, string, CredenialRefObjectType, error). You cannot guess what is what.

Reply:
I've quickly rewritten two `lookup...` methods into one piece of code:

```go
	credID, credSrc, err := reqData.GetCredentialID()
	if err != nil {
		// handle Error
	}

	var tenant string
	var scope string
	switch credSrc {
	case CertificateCredSrc, TokenCredSrc:
		sysAuth, _ := h.systemAuthSvc.GetGlobal(ctx, credID)
		tenant = sysAuth.TenantID
		refObjType := sysAuthToCredRefObjectType(*sysAuth)

		if refObjType == IntegrationSystemRefObject {
			// TODO - handle tenant for integration system
		}
		
		declaredScopes, _ := h.scopesGetter.GetRequiredScopes(buildPath(refObjType))
		scope = strings.Join(declaredScopes, " ")
	case UsernameCredSrc:
		staticUser, _ := h.staticUsersRepo.Get(credID)
		tenant = staticUser.Tenant
		scope = strings.Join(staticUser.Scopes, " ")
	}
```

You can extract it to a separate method and return tenant, scope and error.
I think now it's much simpler. What do you think?